storage/src/property-test/java/tech/pegasys/teku/storage/server/kvstore/serialization/BeaconStateSerializerPropertyTest.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2022 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.storage.server.kvstore.serialization;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static tech.pegasys.teku.storage.server.kvstore.serialization.KvStoreSerializer.createStateSerializer;
+
+import net.jqwik.api.ForAll;
+import net.jqwik.api.Property;
+import net.jqwik.api.constraints.IntRange;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.SpecMilestone;
+import tech.pegasys.teku.spec.TestSpecFactory;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
+import tech.pegasys.teku.spec.networks.Eth2Network;
+import tech.pegasys.teku.spec.util.DataStructureUtil;
+
+public class BeaconStateSerializerPropertyTest {
+  @Property(tries = 10)
+  public void roundTrip(
+      @ForAll final int seed,
+      @ForAll final SpecMilestone milestone,
+      @ForAll final Eth2Network network,
+      @ForAll @IntRange(max = 1000) final int validatorCount) {
+    final Spec spec = TestSpecFactory.create(milestone, network);
+    final DataStructureUtil dataStructureUtil = new DataStructureUtil(seed, spec);
+    final KvStoreSerializer<BeaconState> serializer = createStateSerializer(spec);
+    final BeaconState value = dataStructureUtil.randomBeaconState(validatorCount);
+    final byte[] serialized = serializer.serialize(value);
+    final BeaconState deserialized = serializer.deserialize(serialized);
+    assertThat(deserialized).isEqualTo(value);
```

Comment:
Sadly jqwick's reporting for tests in this style is pretty poor - it doesn't include any of the info about what the parameters that triggered the failure were. While we can get the seed by running gradle with `--info` that also produces a *ton* of other output so it's hard to parse.

Fortunately for cases where there's one fairly simple condition like this we can use a property based test which does give the right info by just returning a boolean. ie:

```java
return deserialised.equals(value);
```

The downside is that we don't get the actual result values printed but at least we can reproduce the problem easily and can then locally modify the test to print whatever other info is useful.  The other option I guess is to use AssertJ's `.describedAs` method to include the parameters in it's exception message but I suspect we'll regularly forget to do that or miss params etc.

Reply:
@ajsutton there's also the [reporter object](https://jqwik.net/docs/current/user-guide.html#platform-reporting-with-reporter-object) and [footnotes](https://jqwik.net/docs/current/user-guide.html#adding-footnotes-to-failure-reports). I will investigate more in a bit.

Reply:
> it doesn't include any of the info about what the parameters that triggered the failure were. While we can get the seed by running gradle with --info that also produces a ton of other output so it's hard to parse.

So I'm not entirely sure what you mean here. If I force a test to fail, I can see the arguments (`arg0`, `arg1`) that resulted in the failure. This is a screenshot of the `./gradlew propertyTest --info` output I see:

<img width="1059" alt="forced_failure_output" src="https://user-images.githubusercontent.com/95511699/168619069-3ea17b37-64d5-44da-83e8-8a6c28bb9c6f.png">

Which is similar to the output in an IDE (like IDEA) but doesn't require the `--info` flag:

<img width="1229" alt="forced_failure_output_idea" src="https://user-images.githubusercontent.com/95511699/168620351-a4a3508f-0f8e-4b9b-8fa1-fab97d43921f.png">

Reply:
So there were three tests that would print a lot to stdout if they failed. I updated those to not use asserts, but rather return a boolean like you mentioned. I wasn't a huge fan of the alternative fixes. I don't think we need to worry about the other tests; their outputs could be a little long, but nothing like printing a beacon state object! What do you think? Is this good enough?

Reply:
I think the issue is more that currently we don't run gradle with the `--info` option and when we do we get a ton of extra logging like:

```
> Task :networking:nat:compileJava UP-TO-DATE
Excluding [io.netty:netty-tcnative-boringssl-static, io.netty:netty-transport-rxtx]
Excluding []
Caching disabled for task ':networking:nat:compileJava' because:
  Build cache is disabled
Skipping task ':networking:nat:compileJava' as it is up-to-date.
:networking:nat:compileJava (Thread[Execution worker for ':' Thread 8,5,main]) completed. Took 0.238 secs.
:networking:nat:classes (Thread[Execution worker for ':' Thread 8,5,main]) started.
```

If we could just enable info logging for the actual property test task that would be fine but we get it enabled for every task along the way as well.

But I'm tempted to just get started like this and we can improve over time. I suspect something that hooks into the life cycle is probably the best long term answer so we can write tests however we like and get the seed in the failure message.